app/controllers/ATSPAYEDataController.scala

```diff
@@ -42,17 +42,25 @@ class ATSPAYEDataController @Inject()(npsService: NpsService, payeAuthAction: Pa
 
   def getATSDataMultipleYears(nino: String, yearFrom: Int, yearTo: Int): Action[AnyContent] = payeAuthAction.async {
     implicit request =>
-      def dataList: Seq[Future[Option[JsValue]]] = (yearFrom to yearTo).toList map { year =>
+      def dataList: Seq[Future[Either[Int, JsValue]]] = (yearFrom to yearTo).toList map { year =>
         callConnector(nino, year) map {
-          case Right(response) => Some(Json.toJson(response))
+          case Right(response) => Right(Json.toJson(response))
           case Left(error) =>
             logger.error(s"Fetching $year data for $nino returned ${error.status}")
-            None
+            Left(error.status)
         }
       }
-      Future.sequence(dataList) map { data =>
-        val flattenedData = data.flatten
-        if (flattenedData.isEmpty) NotFound(s"No data found for $nino") else Ok(Json.toJson(flattenedData))
+
+      Future.sequence(dataList) map { seqEither =>
+        val seqJsValue = seqEither.filter(either => either.isRight) flatMap { right =>
+          right.toOption
+        }
+
+        if (seqJsValue.nonEmpty) Ok(Json.toJson(seqJsValue))
+        else if (seqEither.contains(Left(NOT_FOUND))) NotFound(s"No data found for $nino")
+        else if (seqEither.contains(Left(BAD_REQUEST))) BadRequest(s"Bad request for $nino")
```

Comment:
Similar to the above comment, we are now blocking a user coming into a service if any of their years returns a BadRequest (although this maybe fine), I am not sure if this was your intention.

Reply:
After speaking with Asier I believe this is now the desired functionality. If any of the years return a internal server error or bad request then we show an error page. If we have a mix of Ok and not found, show the ok responses and ignore the not found responses unless there is 100% not found as this means they have no ats.

Reply:
You probably want a foldLeft here. You are going to go over each year and update the result depending on what you have for this year. For this you need a starting value for the result. I would choose an empty list or a Left(NotFound) and then consider each year one by one. Below is with Left(NotFound). What's help when writing a foldLeft is to write the type

You are going to go from List[Either[Error, JsValue]] to Either[Error, List[JsValue]]. In this case only the last or first error is reported depending how you write your foldLeft.

`
seqEither.foldLeft(Left(NotFound): Either[Error, List[JsValue]]) { (result: Either[Error, List[JsValue]], oneYear: Either[Error, JsValue]) => 
  oneYear match {
       // if the current year is not found, just return the result we have so far
       case Left(Notfound) => result
       // if the current is an error, ignore what we have so far and return an error
      case Left(InternalServerError) => Left(InternalServerError)
      // all good, we add the year to the result if possible(3 cases to consider)
     case Right(jsValueOnYear) => {
       result match {
         // the result we have so far already contains an error, then we carry this error ignoring this good year
         case Left(InternalServerError) => Left(InternalServerError)
         // Result is NotFound, no data yet. We return the current year
         case Left(NotFound) => Right(List(jsValueOnYear))
         // there are already something in the result, lets append the new year:
         case Right(jsValueResult) => Right(jsValueResult :: jsValueOnYear) // :: may not be the right operator
      }
    }
 }
`